Refactor the middleware setup [starts v7 breaking changes]

fselmo opened the pull request:
### What was wrong?

In order to process requests and responses separately and to eventually support _batch requests_, we need to uncouple the `make_request()` method from being sandwiched along with all the middleware in each request. If we want APIs that can process request params separately from making a request, we need to refactor quite a bit of the code to massage this sort of logic in.

### How was it fixed?

In this PR I am refactoring the middleware from the functional programming paradigm of a sandwiched request into a class-based structure with request processors and response processors. The ultimate end result for most requests is these classes still wrap the `make_request`, as it is often convenient to be able to replace the `make_request` function in order for a specific request to not be made under certain conditions some middleware necessitate (currently this is only true for the `local_filter_middleware`).

Imo this class-based structure makes the middleware a bit easier to read and create. If a middleware only processes request, it only needs to define a request processor; if the response is the only thing the middleware touches, it can override the response processor on the class, etc. In case the `make_request` method itself needs to be altered, the `_wrap_make_request()` method can be overridden and that would behave similarly to the way the middlewares currently work.

Some things to note:

- This refactor removes the request retry middleware from the provider and removes the concept of provider middleware altogether. Instead, exception retry logic is done as a configuration on the HTTP providers.
- This refactor removes the result generator middlewares and opts for a `request_mocker` fixture for dealing with mocked requests. This has proved to be quite nice to work with, at least for the current tests, and it removes the need to create, inject, and clean up middleware for each test. Instead, the context manager takes care of mocking while inside the manager and cleaning up when exiting the context manager for both sync and async.
- The `simple_cache_middleware` is moved to a decorator on the `make_request` method for the providers. This can be turned on and configured as configurations on the appropriate provider classes.

---

- closes #3021 - Removes the need to pass `w3` into the default middlewares 
- closes #2995 - Remove the need for `abi_middleware` by fixing where the `name_to_address` middleware would override request formatters
- closes #2990 - Removes provider-specific middleware except for a hardcoded specific set of middleware for the in-house `EthereumTesterProvider`
- closes #899 - The current version of the middleware takes `extraData` field and changes it to `proofOfAuthority`. The name was changed from `geth_poa_middleware` to `extradata_to_poa_middleware`

---

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)
- [x] Revise some of the type hinting. In particular the type hinting around the Middleware within manager and provider classes
- Update documentation to reflect above changes ( will be in a separate PR)

#### Cute Animal Picture

![IMG-20240103-WA0005](https://github.com/ethereum/web3.py/assets/3532824/474eb1fb-b4f1-4a1e-8bcc-a24e790f4cdc)

fselmo committed:
Initial commit for middleware refactor

- Refactor middlewares as classes with request processors and response processors
- Apply this refactor to some existing middlewares and test
  - gas_price_strategy_middleware
  - validation / formatting middleware
  - attrdict_middleware
  - (incomplete) ens_name_to_address_middleware
- Refactor the exception_retry_middleware to be a configuration on the http providers

fselmo committed:
Refactor new middleware onion setup

fselmo committed:
Refactor caching as a decorator on ``make_request()``

- Remove all caching middleware.
- Refactor request caching via configurations on the provider classes and decorators on their respective ``make_request()`` methods.

fselmo committed:
Remove middleware dependent on old logic

- This middleware depends on a request always being made when middleware are processed. This isn't how the middleware is processed for websockets and the separation of middleware into request param processing and response processing further helps with being able to support batch requests. Remove all middleware that were dependent on a request always being sandwiched between request processing and response processing.

fselmo committed:
Add request mocker fixture for mocking requests

- Add a ``RequestMocker`` class with ``request_mocker`` pytest fixture to fascilitate request mocking now that the middleware doesn't sequester the ``make_request()`` call. Since we can't swap the ``make_request()`` on the fly, mock results and errors using this test utility class.
- Linting
- Some cleanup along the way

fselmo committed:
Some refactoring to make cached_requests work with WSV2

fselmo committed:
WIP: Fix or comment out core test errors so core suite can run

fselmo committed:
Allow for manipulating the method via middleware

- In the middleware refactor, we weren't accounting for manipulating the ``method``, only the params. Allow for this kind of manipulation as it is necessary for ``sign_and_send_raw_middleware``, for example.

fselmo committed:
Re-introduce the sandwiched middleware model via a method on the middleware class

- Use a repr() for the middleware onion name when a non-hashable, such as the curried `build` method on a web3middlewarebuilder is used.

fselmo committed:
Build on the wrapped make_request middleware refactor

- Add support back for ``local_filter_middleware`` by overriding the ``_wrap_make_request()`` of the base middleware class. If this method is overridden, the ``make_request`` can be effectively replaced if a request is not needed.

fselmo committed:
Fix remaining tests that were broken due to refactor

fselmo committed:
Remove the necessity for the ``abi_middleware``

- closes #2995
- ``ens_name_to_address_middleware`` would return the ``bytes`` address, rather than checksummed address. Since this happens after the ``request_formatters`` have been applied, this would override the results from the ``ABI_REQUEST_FORMATTERS`` which take ``address`` types and format them to a checksummed hex address.

fselmo committed:
Reinstate the http retry request tests as configuration

fselmo committed:
Fix type hinting for refactor

fselmo committed:
Fix inconsistency with eth-tester integration test

fselmo committed:
Remove unnecessary imports hidden by # noqa: F401

fselmo committed:
use request mocker + no need to format logsBloom in eth-tester middleware

- Use request_mocker where we can
- result_processors take care of the formatting... this was causing the ``None`` value for logsBloom to raise when mocked

fselmo committed:
Rename all references of geth_poa to extradata_to_poa

- The middleware changed names in the refactor, but this commit puts the finishing touches and closes #899

fselmo committed:
Minor cleanups from refactor PR

reedsa commented on tests/core/contracts/test_contract_example.py:
```diff
@@ -121,10 +119,9 @@ def async_eth_tester():
-    w3 = Web3(provider, modules={"eth": [AsyncEth]}, middlewares=provider.middlewares)
-    w3.eth.default_account = await w3.eth.coinbase
-    return w3
+    async_w3 = AsyncWeb3(AsyncEthereumTesterProvider())
```
Very nice 😻

reedsa commented on tests/core/manager/conftest.py:
```diff
@@ -15,15 +19,19 @@ class Wrapper:
+                    def _wrap_make_request(self, make_request):
+                        def middleware_fn(method, params):
+                            params.append(key)
+                            method = "|".join((method, key))
```
Is this pipe a delimiter for the request so we have a way to determine each unique method call?

fselmo replied:
Yeah, so we have a way of keeping track of unique middleware "keys" and asserting the order they wrap things in. I think it's really only used in [this test](https://github.com/ethereum/web3.py/blob/0fc0e064fbba6eca7a78b044dc8bc5888d5a0c45/tests/core/manager/test_provider_request_wrapping.py#L29-L29) if I'm searching correctly. I didn't change the logic here at all though, this code is from 2017 it looks like.

fselmo committed:
'name_to_address' -> 'ens_name_to_address' for consistency.

fselmo committed:
Make caching more robust; add tests back for cached requests

- Make request caching more robust by re-introducing the lock from the middleware as well as checking for "error" and None responses and not caching those.
- Re-introduce the tests for the simple_cache_middleware as caching utils tests.

[Unrelated]

- Keep the decorator logic on ``make_request`` when using the ``request_mocker`` fixture so we can still utilized the request mocker for testing cached requests.

pacrob commented on tests/core/caching-utils/test_request_caching.py:
```diff
@@ -0,0 +1,238 @@
+        assert err_a.value.args != err_b.value.args
+
+
+def test_request_caching_does_not_cache_endpoints_not_in_whitelist(w3):
```
```suggestion
def test_request_caching_does_not_cache_endpoints_not_in_allowlist(w3):
```

pacrob commented on tests/core/caching-utils/test_request_caching.py:
```diff
@@ -0,0 +1,238 @@
+
+
+@pytest.mark.asyncio
+async def test_async_request_caching_does_not_cache_non_whitelist_endpoints(
```
```suggestion
async def test_async_request_caching_does_not_cache_non_allowlist_endpoints(
```

kclowes commented on docs/middleware.rst:
```diff
@@ -446,16 +446,15 @@ Time-based Cache Middleware
 .. note::
     It's important to inject the middleware at the 0th layer of the middleware onion:
-    ``w3.middleware_onion.inject(geth_poa_middleware, layer=0)``
+    ``w3.middleware_onion.inject(extradata_to_poa_middleware, layer=0)``
```
this is a nit you can definitely leave, but maybe call this: `poa_extradata_middleware`?

fselmo replied:
My thought with `extradata_to_poa_middleware` is that it tells you exactly what the middleware does. I may leave as is but I'm not tied to it if you have a strong opinion here. Let me know.

kclowes replied:
Nope, no strong opinion!

kclowes commented on web3/_utils/module_testing/utils.py:
```diff
@@ -0,0 +1,194 @@
+        return self
+
+    # define __exit__ with typing information
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
```
nit: Can we tighten up any of these types?

fselmo replied:
That comment was supposed to be a `# TODO`, thanks for catching 👍🏼

kclowes commented on tests/core/caching-utils/test_request_caching.py:
```diff
@@ -0,0 +1,238 @@
+                result_b = w3_b.manager.request_blocking("eth_chainId", [])
+                result_c = w3_c.manager.request_blocking("eth_chainId", [])
+
+    assert result_a != result_b != result_c
```
nit: this is redundant if we're already checking what the actual values are in the next couple lines 🤷

pacrob commented on web3/middleware/exception_retry_request.py:
It looks like this file was emptied but not deleted. Any reason it shouldn't be?

pacrob commented on web3/middleware/filter.py:
```diff
@@ -411,9 +373,9 @@ async def async_iter_latest_block(
     while True:
-        latest_block = await w3.eth.block_number  # type: ignore
+        latest_block = await w3.eth.block_number
         # type ignored b/c is_bounded_range prevents unsupported comparison
```
can remove `type ignored b/c` because `type: ignore` removed

fselmo replied:
This is separate. It's talking about the still-type-ignored `is_bounded_range` line below the comment.

pacrob commented on web3/middleware/cache.py:
empty file?

kclowes commented on web3/main.py:
```diff
@@ -472,8 +469,8 @@ async def is_connected(self, show_traceback: bool = False) -> bool:
     @property
-    def middleware_onion(self) -> AsyncMiddlewareOnion:
-        return cast(AsyncMiddlewareOnion, self.manager.middleware_onion)
+    def middleware_onion(self) -> MiddlewareOnion:
```
nit: we could probably put this on `BaseWeb3` now?

pacrob commented on tests/core/providers/test_http_request_retry.py:
```diff
@@ -0,0 +1,183 @@
+    assert make_post_request_mock.call_count == TEST_RETRY_COUNT
+
+
+def test_exception_retry_config_is_strictly_on_http_provider():
```
Could do WS, and AsyncWeb3 versions too, for completeness. But a take/leave as you like.

fselmo replied:
Yeah I just ported the test as it was over here. Refraining from adding too many additional things at the moment but this seems simple enough. I'll put it in 👍🏼

kclowes commented on web3/middleware/base.py:
```diff
@@ -0,0 +1,129 @@
+# --- type definitions --- #
+
+Middleware = Type[Web3Middleware]
+MiddlewareOnion = NamedElementOnion[str, Middleware]
```
nit: is this `str` accurate any more?

fselmo replied:
All keys (`TKey` names) in the onion should still be strings, yeah. That logic hasn't changed other than to use the `repr()` of the element if the `TValue` isn't hashable.

fselmo replied:
Actually, I changed that due to [this comment](https://github.com/ethereum/web3.py/blob/45db990ec8240930a11a38e467bfcf200e46aa50/web3/datastructures.py#L32-L39). If `TKey` is bound to a hashable we should define a `__hash__()` that would reflect any changes to the middleware class.

fselmo replied:
False alarm, this was mostly because we borrow the defined ``TKey`` for ``AttributeDict`` data structure. We can revisit this later but we already break these rules in the middleware onion when we set a ``TKey`` as a "name" string. I took out that last commit, adding a newsfragment and merging!

fselmo committed:
Changes from comments on PR #3169

pacrob reviewed (approved):
Nicely done! :weight_lifting:
GTG for me.

kclowes commented on web3/providers/async_base.py:
```diff
@@ -40,53 +48,61 @@
     ) -> Callable[..., Coroutine[Any, Any, RPCResponse]]:
         # type ignored b/c tuple(MiddlewareOnion) converts to tuple of middlewares
-        all_middlewares: Tuple[AsyncMiddleware] = tuple(outer_middlewares) + tuple(self.middlewares)  # type: ignore  # noqa: E501
+        middlewares: Tuple[Middleware, ...] = tuple(middlewares)  # type: ignore
```
nit: does this `type: ignore` still need to be here?

fselmo replied:
Unfortunately yeah. I played around with this a bit but didn't get too far. It comes fromt he fact that the `__iter__()` method defined before my time will turn the onion into a `Tuple[Middleware, ...]` but to type it as such isn't straightforward. The type here _is_ correct, we just don't have a very elegant solution for this (yet) and maybe we can tackle that as a separate issue altogether. Could be worth looking into. I think it has to do with satisfying the `Mapping` inheritance for the `NamedElementOnion`.

fselmo replied:
Just kidding. The moment I said a definitive "yeah we have to keep it" something clicked. It was easily solved by creating a method to tupelize the onion into a tuple of middlewares with the correct type hinting. I pushed that up as a separate commit in case that isn't ideal but I prefer than to the `type: ignore`s myself. Let me know what you think on that.

kclowes replied:
Nice! Yeah, I think I prefer that too

kclowes commented on web3/providers/base.py:
```diff
@@ -47,44 +72,42 @@ class BaseProvider:
-                self._generate_request_func(w3, all_middlewares),
+                middlewares,
+                combine_middlewares(
+                    middlewares=middlewares,  # type: ignore
```
why do we need to ignore the type here? might be worth a comment?

fselmo replied:
This comment actually made me solve the comment above. All these `type: ignores` go away with the `as_tuple_of_middlewares()` method. I could use help naming it though 😆. I like how explicit that name is and I felt like `tupleize` just doesn't explain what it does for a reader trying to move on past the code. Any thoughts there would be appreciated.

kclowes replied:
I can't think of anything better off the top of my head but will let you know if I do!

kclowes reviewed (approved):
Left a few nits to take or leave, but looks good!

Nov1kov commented:
Sorry for an almost off-topic. But why the package [eth_tester](https://github.com/ethereum/web3.py/tree/main/web3/providers/eth_tester) couldn't be fully in the [eth-tester pypi](https://github.com/ethereum/eth-tester)?

For adding implementation for one method in the eth-tester, I have to implement formatter also in web3.py

fselmo committed:
Use as_tuple_of_middlewares() to get around typing issues

- Use an internal method for the ``NamedMiddlewareOnion`` to generate
  the tuple of middlewares, instead of relying on ``tuple()``.

fselmo commented:
@kclowes I think I addressed all your comments. Let me know if there's anything else outstanding and how you feel about the [tupleization commit](a548a4e420b304d548270657b2868cb0be8ee6d6).

kclowes commented:
LGTM! 🚀

fselmo committed:
newsfragment for 3169
